modules/fhir-structure/src/blaze/fhir/spec.clj

```diff
@@ -54,64 +56,82 @@
+  "Given a vector `path` and some `data` structure containing maps and vectors,
+  for which the entries in elem correspond to the keys and vector entries of the
+  given data, returns a vector of the same length as path with all vector entries
+  from data replaced by their indices."
+  {:arglists '([path data])}
   [[elem & elems] data]
-  (if (and (map? elem) (vector? data))
+  (if (vector? data)
     (cons (first (keep-indexed #(when (= %2 elem) %1) data))
           (when elems (fhir-path-data elems elem)))
     (cons elem (if elems (fhir-path-data elems (get data elem)) []))))
 
 
 (defn fhir-path
-  "Given a vector `path` and a `resource`, returns the fhir-path as String for
+  "Given a vector `path` and a `resource`, returns the fhir-path as string for
   the given path."
   [path resource]
-  (subs (reduce (fn [x y] (str x (if (keyword? y)
-                                   (format ".%s" (name y))
-                                   (format "[%s]" y)))) ""
-                (fhir-path-data path resource)) 1))
+  (->>(fhir-path-data path resource)
+      (reduce (fn [list elem] (if (keyword? elem)
+                                (conj list (name elem))
+                                (update list
+                                        (dec (count list))
+                                        str
+                                        (format "[%d]" elem))))
+              [])
+      (str/join ".")))
 
 
 (defn get-regex
   "Returns the first regex pattern in `form`."
   [form]
-  (clojure.walk/postwalk
+  (walk/postwalk
     #(if (sequential? %)
        (some identity %)
        (when (= (type %) Pattern) %))
     form))
 
 
-(defn get-diagnostics-from-problem
-  "Returns a diagnostics message from the given arguments pred, val and via.
-  pred: A Function to evaluate the given val.
-  val: Error Value.
-  via: Vector-path of fhir types to the given error.
-  If pred contains Regex-patetrns to check, the first is included in diagnostics
-  message."
+(defn- extended-fhir-namespace?
+  "Returns the namesake part and name of `key` if the namespace of `key` is more
+  than :fhir. Otherwise returns nil."
+  [key]
+  (when (str/starts-with?  (str key) ":fhir")
+    (str/replace (subs (str key) 6) "/" ".")))
+
+
+(defn- get-diagnostics-from-problem
+  "Returns a diagnostics message from the given arguments `pred`, `val` and `via`.
+  `pred`: A Function to evaluate the given `val`.
+  `val`: Error Value.
+  `via`: Vector-path of fhir types to the given error.
+  If `pred` contains regex-patterns to check, the first one is included in
+  diagnostics message."
   [pred val via]
   (if-let [regex (get-regex pred)]
-    (format "Error on Value `%s`. Expected type is `%s`, regex `%s`."
+    (format "Error on value `%s`. Expected type is `%s`, regex `%s`."
             val (name (last via)) regex)
-    (format "Error on Value `%s`. Expected type is `%s`."
-            val (name (last via)))))
+    (if (= (str pred)  "clojure.core/coll?")
```

Comment:
(= `coll? pred)

Pred is a symbol and you can write a literal symbol in source code by quoting it. the back tics syntax quote will automatically resolve the symbol according the namespace declarations.